x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -7,22 +7,19 @@ import (
 	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
 
 	"github.com/cosmos/interchain-security/v4/x/ccv/provider/types"
+	"sort"
 )
 
+// HandleOptIn prepares validator `providerAddr` to opt in to `chainID` with an optional `consumerKey` consumer public key.
+// Note that the validator only opts in at the end of an epoch.
 func (k Keeper) HandleOptIn(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress, consumerKey *string) error {
 	if !k.IsConsumerProposedOrRegistered(ctx, chainID) {
 		return errorsmod.Wrapf(
 			types.ErrUnknownConsumerChainId,
 			"opting in to an unknown consumer chain, with id: %s", chainID)
 	}
 
-	if k.IsToBeOptedOut(ctx, chainID, providerAddr) {
-		// a validator to be opted in cancels out with a validator to be opted out
-		k.DeleteToBeOptedOut(ctx, chainID, providerAddr)
-	} else if !k.IsToBeOptedIn(ctx, chainID, providerAddr) && !k.IsOptedIn(ctx, chainID, providerAddr) {
-		// a validator can only be set for opt in if it is not opted in and not already set for opt in
-		k.SetToBeOptedIn(ctx, chainID, providerAddr)
-	}
+	k.SetOptedIn(ctx, chainID, providerAddr)
```

Comment:
just to make sure: if this function returns an error, this SetOptedIn is rolled back, right? Otherwise, it could happen that someone tries to opt in, key assignment fails, but this still opts them in

Reply:
As far as I know, the execution is performed on cached context and state would only be committed if all the messages of a transactions are successful. See [here](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/baseapp.go#L814):
```
// [...]. All state transitions occur through
// a cached Context depending on the mode provided. State only gets persisted
// if all messages get executed successfully and the execution mode is DeliverTx.
```

Reply:
Ok, yeah that matches my understanding